rust/src/transformations/sum/mod.rs

```diff
@@ -101,202 +71,182 @@ where
+
+                    } else {
+                        // 2. fall back to ordered summation
+                        make_sized_bounded_float_ordered_sum::<Pairwise<_>>(size, bounds)
+                    }
                 } else {
-                    // 2. fall back to ordered summation
-                    make_sized_bounded_float_ordered_sum::<Pairwise<_>>(size, bounds)
+                    if !Pairwise::<Self>::float_sum_can_overflow(DEFAULT_SIZE_LIMIT, bounds)? {
+                        // 1. if the sum can't overflow,
+                        //    then do a no-op unordering and use a more computationally efficient sum without saturation arithmetic
+                        let domain = VectorDomain::new(AtomDomain::new_closed(bounds.clone())?);
+                        make_unordered(domain)? >> make_bounded_float_checked_sum::<Pairwise<_>>(DEFAULT_SIZE_LIMIT, bounds)?
+
+                    } else {
+                        // 2. sum can overflow, but data is already sorted, so use a saturating sum
+                        make_bounded_float_ordered_sum::<Pairwise<_>>(DEFAULT_SIZE_LIMIT, bounds)
+                    }
                 }
             }
         })+
     };
 }
-impl_make_sized_bounded_sum_float! { f32 f64 }
+impl_make_sum_float! { f32 f64 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
 
     #[test]
-    fn test_make_bounded_sum_l1() {
-        let transformation = make_bounded_sum::<SymmetricDistance, i32>((0, 10)).unwrap_test();
-        let arg = vec![1, 2, 3, 4, 5];
-        let ret = transformation.invoke(&arg).unwrap_test();
-        let expected = 15;
-        assert_eq!(ret, expected);
-    }
-
-    #[test]
-    fn test_make_bounded_sum_l2() {
-        let transformation = make_bounded_sum::<SymmetricDistance, i32>((0, 10)).unwrap_test();
+    fn test_make_sum_l1() -> Fallible<()> {
+        let input_space = (
+            VectorDomain::new(AtomDomain::new_closed((0, 10))?),
+            SymmetricDistance::default(),
+        );
+        let transformation = (input_space >> then_sum())?;
         let arg = vec![1, 2, 3, 4, 5];
         let ret = transformation.invoke(&arg).unwrap_test();
         let expected = 15;
         assert_eq!(ret, expected);
+        Ok(())
     }
 
     #[test]
-    fn test_make_sized_bounded_sum() {
-        let transformation =
-            make_sized_bounded_sum::<SymmetricDistance, i32>(5, (0, 10)).unwrap_test();
+    fn test_make_sized_bounded_sum() -> Fallible<()> {
+        let input_space = (
+            VectorDomain::new(AtomDomain::new_closed((0, 10))?).with_size(5),
+            SymmetricDistance::default(),
+        );
+        let transformation = (input_space >> then_sum())?;
         let arg = vec![1, 2, 3, 4, 5];
         let ret = transformation.invoke(&arg).unwrap_test();
         let expected = 15;
         assert_eq!(ret, expected);
+        Ok(())
```

Comment:
Given the separate trait impls, it'd be nice to have tests at least for InsertDeleteDistance vs SymmetricDistance and float vs int.